src/core/infer_request.cc

```diff
@@ -789,6 +805,29 @@ InferenceRequest::Input::AppendData(
   return Status::Success;
 }
 
+Status
+InferenceRequest::Input::AppendDataForDevice(
+    const void* base, size_t byte_size, TRITONSERVER_MemoryType memory_type,
+    int64_t memory_type_id, TRITONSERVER_InstanceGroupKind device_kind,
+    int device_id)
+{
+  DeviceInfo dev_info(device_kind, device_id);
+  auto device_data = device_specific_data_.find(dev_info);
+  has_device_specific_data_ = true;
+  if (device_data == device_specific_data_.end()) {
+    device_specific_data_.insert(std::make_pair(dev_info, new MemoryReference));
+  }
+  device_data = device_specific_data_.find(dev_info);
```

Comment:
`insert()` returns the iterator to the new element so you don't need to find again.